document developer workflow (#888)
digital-asset/docs.daml.com: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst

upgrade that dependency to a new version.

For this reason, it is :ref:`strongly recommended that interfaces always be defined
in their own packages separately from templates <upgrades-separate-interfaces-and-exceptions>`.

Developer Workflow
------------------

> The following recommendations about how to set up your development environment
> for SCU will help you iterate more quickly on your projects.

Reviewer: ```suggestion
This section contains suggestions on how to set up your development environment
for SCU to help you iterate more quickly on your projects.
```

Multi-Package Builds
~~~~~~~~~~~~~~~~~~~~

> Following the :ref:`best practices <upgrades-best-practices>` outlined below in this
> document as well as the :ref:`testing recommendations <upgrades-testing>` will lead to a
> proliferation of packages in your project.  Use :ref:`Multi-Package builds
> <multi-package-build>` to reliably rebuild these packages as you iterate on your
> project. Multi-package builds will also enable cross-package navigation in Daml
> Studio.

Reviewer: ```suggestion
Following the :ref:`best practices <upgrades-best-practices>` outlined below and the :ref:`testing recommendations <upgrades-testing>` leads to a
proliferation of packages in your project.  Use :ref:`Multi-Package builds
<multi-package-build>` to reliably rebuild these packages as you iterate on your
project. Multi-package builds also enable cross-package navigation in Daml
Studio.
```

> The :ref:`Multi-package builds for upgrades <upgrades-multi-package>` section
> goes into more detail on how to set up multi-package builds for SCU and how it 
> can help with testing in particular.

Reviewer: ```suggestion
The :ref:`Multi-package builds for upgrades <upgrades-multi-package>` section
goes into more detail on how to set up multi-package builds for SCU and how it 
can help with testing.
```

Working with a Running Canton Instance
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> With SCU, it is no longer possible to iterate on a package by
> uploading it to a running participant after each rebuild. This is because
> a participant will reject two packages with the same name and version whose content
> differs.

Reviewer: ```suggestion
With SCU, it is no longer possible to iterate on a package by
uploading it to a running participant after each rebuild. A participant rejects two packages with the same name and version whose content
differs.
```

There are two ways to work around this:

>   - restart the participant after each rebuild;
>   - change the version name of the package before each rebuild.

Reviewer: ```suggestion
  - restart the participant after each rebuild
  - change the version name of the package before each rebuild
```

:ref:`Environment variable interpolation <environment-variable-interpolation>`
in the Daml Assistant can help with the latter. In the ``daml.yaml`` file of each
of your packages, append an environment variable to the name of the package:

.. code:: daml

    name: my-package-v${UNIQUE_BUILD_ID}

Make sure to also append the variable to the name of the DAR file produced by
``my-package`` in the ``daml.yaml`` files that depend on it:

.. code:: daml

  dependencies:
  - daml-prim
  - daml-stdlib
  data-dependencies:
  - ../my-package/.daml/dist/my-package-${UNIQUE_BUILD_ID}.dar

Then, before invoking ``daml build --all``, increment the ``UNIQUE_BUILD_ID``

> environment variable. This will ensure that the DAR files produced by the build
> are unique and can be uploaded to the participant without conflict.

Reviewer: ```suggestion
environment variable. This ensures that the build produces unique DAR files that can be uploaded to the participant without conflict.
```

Working with the Daml Sandbox
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For the same reason, the :ref:`Daml sandbox <sandbox-manual>` does not support
hot-swapping of SCU-enabled (LF 1.17) packages.

There are two ways to work around this:

>   - restart the sandbox after each rebuild;
>   - change the version name of your packages after each rebuild, as outlined in
>     the previous section.

Reviewer: ```suggestion
  - restart the sandbox after each rebuild
  - change the version name of your packages after each rebuild, as outlined in
    the previous section
```

Studio.

The :ref:`Multi-package builds for upgrades <multi_package_upgrades>` section
goes into more detail on how to set up multi-package builds for SCU and how it 
can help with testing in particular.

Working with a Running Canton Instance
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

With SCU, it is no longer possible to iterate on a package by
uploading it to a running participant after each rebuild. This is because
a participant will reject two packages with the same name and version whose content
differs.

There are two ways to work around this:
  - restart the participant after each rebuild;
  - change the version name of the package before each rebuild.

:ref:`Environment variable interpolation <environment-variable-interpolation>`
in the Daml Assistant can help with the latter. In the ``daml.yaml`` file of each
of your packages, append an environment variable to the name of the package:

.. code:: daml

    name: my-package-v${UNIQUE_BUILD_ID}

Make sure to also append the variable to the name of the DAR file produced by
``my-package`` in the ``daml.yaml`` files that depend on it:

.. code:: daml

  dependencies:
  - daml-prim
  - daml-stdlib

>   - '../my-package/.daml/dist/my-package-${UNIQUE_BUILD_ID}.dar'

Reviewer 2: We don't use quotes anywhere else in the doc. Also I would prefer it to be a data-dependency, but I'm easily convinced otherwise

> Then, before invoking ``daml build --all``, increment the ``UNIQUE_BUILD_ID``

Reviewer 2: Could you verify this works as well? I'm worried the caching won't notice this change

Reviewer 2: Have verified

  build-options:
  - --target=1.17

> In order to manage the complexity of working with multiple packages at once, it
> is recommended to use :ref:`multi-package builds <multi-package-build>`.

Reviewer: ```suggestion
To manage the complexity of working with multiple packages at once, we recommend using :ref:`multi-package builds <multi-package-build>`.
```

**Note:** We still recommend against depending on Daml Script for
ledger-uploaded packages, even in this case with non-serializable positions.

> .. _testing:

Reviewer 2: These labels are docs wide, I'd call it something like `upgrades-testing`. Same for the `best_practices`

Reviewer 2: (I believe we also prefer dashes to underscores, other than ofc the one at the start. I'd check some others to see if thats true)

Reviewer 2: I've updated these two, along with a bunch of other ones in the doc
